core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -65,12 +65,53 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
   /* the set of logs currently being cleaned */
   private val inProgress = mutable.HashMap[TopicPartition, LogCleaningState]()
 
+  /* the set of uncleanable partitions (partitions that have raised an unexpected error during cleaning)
+   *   for each log directory */
+  private val uncleanablePartitions = mutable.HashMap[String, mutable.Set[TopicPartition]]()
+
+  /* the set of directories marked as uncleanable and therefore offline */
+  private val uncleanableDirs = mutable.HashSet[String]()
+
   /* a global lock used to control all access to the in-progress set and the offset checkpoints */
   private val lock = new ReentrantLock
 
   /* for coordinating the pausing and the cleaning of a partition */
   private val pausedCleaningCond = lock.newCondition()
 
+  /* gauges for tracking the number of partitions marked as uncleanable for each log directory */
+  for (dir <- logDirs) {
+    newGauge(
+      "uncleanable-partitions-count",
+      new Gauge[Int] { def value = uncleanablePartitions.get(dir.getAbsolutePath).map(_.size).getOrElse(0) },
```

Comment:
Is this called from another thread? If so, I am not sure if this is thread-safe. Might be worth verifying. Same for the uncleanable bytes computation below.

Reply:
Yes, I haven't addressed this comment yet. I remember I thought about it while implementing it but thought it was okay for some reason. Will get back to you in soon